src/bluesearch/entrypoint/database/run.py

```diff
@@ -0,0 +1,636 @@
+        """Define conditional dependencies."""
+        if self.source == "arxiv":
+            return self.clone(ConvertPDFTask)
+        else:
+            return self.clone(PerformFilteringTask)
+
+    def output(self) -> luigi.LocalTarget:
+        """Define output folder."""
+        output_file = Path(self.input().path).parent / "parsed"
+
+        return luigi.LocalTarget(str(output_file))
+
+    def program_args(self) -> list[str]:
+        """Define subprocess arguments."""
+        output_dir = Path(self.output().path)
+
+        if (output_dir.parent / "converted_pdfs").exists():
+            input_dir = output_dir.parent / "converted_pdfs"
+        else:
+            input_dir = output_dir.parent / "filtered"
+
+        # Determine parser
+        source2parser = {
+            "arxiv": "tei-xml-arxiv",
+            "biorxiv": "jats-meca",
+            "medrxiv": "jats-meca",
+            "pmc": "jats-xml",
+            "pubmed": "pubmed-xml-set",
+        }
+        parser = source2parser[self.source]
+
+        command = [
+            *BBS_BINARY,
+            "parse",
+            *VERBOSITY,
+            parser,
+            str(input_dir),
+            str(output_dir),
+        ]
+
+        return command
+
+
+@requires(ParseTask)
+class AddTask(ExternalProgramTask):
+    """Add parsed articles to the database.
+
+    This step is considered done if all articles inside of `parsed/` are
+    already in the database.
+    """
+
+    db_url = luigi.Parameter()
+    db_type = luigi.Parameter()
+
+    def complete(self) -> bool:
+        """Check if all articles inside of `parsed/` are in the database."""
+        # If all the articles are inside
+        if self.db_type == "sqlite":
+            prefix = "sqlite:///"
+        elif self.db_type == "postgres":
+            prefix = "postgresql+pg8000://"
+        else:
+            raise ValueError
+
+        engine = sqlalchemy.create_engine(f"{prefix}{self.db_url}")
+
+        input_dir = Path(self.input().path)
+        if not input_dir.exists():
+            return False
+
+        all_uids = [
+            article.stem for article in input_dir.iterdir() if article.suffix == ".json"
+        ]
+
+        new_uids = []
+        for uid in all_uids:
+            query = sqlalchemy.text(
+                "SELECT article_id from articles WHERE article_id = :uid"
+            )
+            res = engine.execute(query, uid=uid).fetchall()
+
+            if not res:
+                new_uids.append(uid)
+
+        return not new_uids
+
+    def program_args(self) -> list[str]:
+        """Define subprocess arguments."""
+        input_dir = Path(self.input().path)
+
+        command = [
+            *BBS_BINARY,
+            "add",
+            *VERBOSITY,
+            self.db_url,
+            input_dir,
+            f"--db-type={self.db_type}",
+        ]
+
+        return command
+
+
+def get_all_dependencies(task: luigi.Task) -> set[luigi.Task]:
```

Comment:
If I understand correctly, you are using this function only to retrieve the class name of the `final_task`?
But then wouldn't it be easier to just do something like `getattr` to get the class object from its `str` name?

Reply:
> No, this function gets all task instances.

Sure, this was clear. I was referring to what the actual usage of it in the code – namely the fact that "we only select the one that the user specified as `--final-task`".


----

> The approach your suggested would require passing all the parameters necessary to instantiate a given class.

Ah now I see the issue, thanks! Anyway, notice that `luigi` allows to provide the `Parameter` values of any `Task`  in the `luigi.cfg`. This means that we can simply execute 
```python
# Note: I am calling __init__() w/o defining values for the parameters here!!
luigi.build(tasks=[AddTask()])  
```
as long as we have a `luigi.cfg` that looks like this one:
```
[DownloadTask]
source = arxiv
from_month = 2021-12
...

[GenerateWordsTask]
...

[AddTask]
...
```

Now, there is only one small issue: since you have used `@inherits` and `@requires` everywhere, `luigi` expects **every** `Task` we have defined in our pipeline to have those parameters set somewhere. If we use the aforementioned approach with the parameters set in `luigi.cfg`, we have two possibilities.
1. If `AddTask` is the `final_task`, then we can just define all parameters under `[AddTask]`.
2. If the `final_task` is something else that we don't know in advance (our case), then we need every `[XXXTask]` to have defined all the inherited parameters. For instance, every `[XXXTask]` will need to define `source = ...`!

I think that our use case is option 2., but the fact of having to define `source = ...` many times in the `luigi.cfg` is not really great. Anyway, I think that this is not really `luigi`'s fault, it's more the fact we have those `@inherits` and `@requires` everywhere, even though I am not sure that we should really do that — why setting `@requires(ParseTask)` for `AddTask`, as the latter does not need any of the parameters of the former?

Reply:
Sorry for the lengthy comment, here's a summary of the alternative I see for the current approach.

➡️ (Unless we really need to re-use parameters in downstream tasks) get rid of the `@inherits` and `@requires` + set `Parameter` values of each `Task` in the corresponding section in the `luigi.cfg`.

And here's the main benefits.

😊 Use a configuration file to configure the parameters of each `luigi` `Task` seems to be cleaner. Also, no need to crawl the `Task` dependency graph to retrieve the various instances. Finally, by getting rid of unneeded `@inherits`/`@requires`, each `Task` only has attached to it the `Parameter`s it really needs.

Reply:
We tested it, and we can indeed use empty `__init__()` for any `Task`, as long as the `luigi.cfg` defines values for each required `Parameter`.

Still to do:
- [ ] what happens if the `luigi.cfg` is modified _after_ the `luigi.build()` is called? is it re-loaded? or ignored? (especially after it starts a new task)
- [ ] is there any way to specify only once the value of `source` which is shared by many `Task`s? e.g. maybe in the `luigi.cfg`? or as a dummy `Task`?

Reply:
From my experiments today, it seems:
- The `luigi.cfg` configuration file is not reloaded during a run. It means that even if the file is modified while luigi pipeline is running, the initial parameters are kept.
- The way to handle shared parameters is through a `luigi.Config` class containing the global parameters. See (here)[https://stackoverflow.com/questions/48509083/how-to-make-a-parameter-available-to-all-luigi-tasks]